src/main/java/techreborn/blockentity/machine/misc/PumpBlockEntity.java

```diff
@@ -0,0 +1,211 @@
+	}
+
+	public void setRange(int newRange) {
+		range = newRange;
+		if (range < 1) range = 1;
+		if (range > 50) range = 50;
+	}
+
+	int depth = 10;
+
+	public int getDepth() {
+		return depth;
+	}
+
+	public void setDepth(int newDepth) {
+		depth = newDepth;
+		if (depth < 1) depth = 1;
+		if (depth > 50) depth = 50;
+	}
+
+	public PumpBlockEntity(BlockPos pos, BlockState state) {
+		super(TRBlockEntities.PUMP, pos, state, "Pump", TechRebornConfig.pumpMaxInput, TechRebornConfig.pumpMaxEnergy, TRContent.Machine.PUMP.block, 0);
+		this.inventory = new RebornInventory<>(1, "PumpBlockEntity", 64, this);
+		this.tank = new Tank("PumpBlockEntity", TANK_CAPACITY, this);
+		this.exhausted = false;
+	}
+
+	@Override
+	public Tank getTank() {
+		return tank;
+	}
+
+	@Override
+	public ItemStack getToolDrop(PlayerEntity p0) {
+		return TRContent.Machine.PUMP.getStack();
+	}
+
+	@Override
+	public void readNbt(final NbtCompound tagCompound) {
+		super.readNbt(tagCompound);
+		tank.read(tagCompound);
+		if (tagCompound.contains("rangeConfig")) {
+			NbtCompound rangeConfig = tagCompound.getCompound("rangeConfig");
+			this.range = rangeConfig.getInt("range");
+			this.depth = rangeConfig.getInt("depth");
+		} else {
+			this.range = DEFAULT_RANGE;
+			this.depth = DEFAULT_DEPTH;
+		}
+	}
+
+	@Override
+	public void writeNbt(final NbtCompound tagCompound) {
+		super.writeNbt(tagCompound);
+		tank.write(tagCompound);
+		NbtCompound rangeConfig = new NbtCompound();
+		rangeConfig.putInt("range", range);
+		rangeConfig.putInt("depth", depth);
+		tagCompound.put("rangeConfig", rangeConfig);
+	}
+
+	@Override
+	public BuiltScreenHandler createScreenHandler(int syncID, PlayerEntity player) {
+		return new ScreenHandlerBuilder("pump")
+			.player(player.getInventory()).inventory().hotbar().addInventory()
+			.blockEntity(this)
+			.energySlot(0, 8, 72)
+			.sync(tank)
+			.syncEnergyValue()
+			.sync(this::getDepth, this::setDepth)
+			.sync(this::getRange, this::setRange)
+			.addInventory()
+			.create(this, syncID);
+	}
+
+	@Override
+	public void tick(World world, BlockPos pos, BlockState state, MachineBaseBlockEntity blockEntity) {
+		super.tick(world, pos, state, blockEntity);
+
+		if (world == null || world.isClient) return;
+
+		//do nothing if all liquids have been exhausted
+		if (this.exhausted) return;
+
+		//if pumping time completed
+		if (pumpedTargetBlockPos != null) {
+			if ((world.getTime() - timePumpingStarted >= (long) (TechRebornConfig.pumpTicksToComplete * (1 - getSpeedMultiplier())))
+				&& getEnergy() >= (long) (TechRebornConfig.pumpEnergyToCollect * getPowerMultiplier())) {
+				//collect fluid
+				Fluid fluid;
+				FluidValue fluidValue = FluidValue.BUCKET;
+				SoundEvent soundEvent;
+				BlockState blockState = world.getBlockState(pumpedTargetBlockPos);
+				if (blockState.getBlock() == Blocks.WATER) {
+					fluid = Fluids.WATER;
+					soundEvent = SoundEvents.ITEM_BUCKET_EMPTY;
+				} else if (blockState.getBlock() == Blocks.LAVA) {
```

Comment:
I would make the first check for lava and then the default water, since theoretically we could also pump other fluids. For example we have generating oil lakes.

Reply:
I can swap the checks, but the pump is currently limited to water and lava anyway, I tried to pump oil from an artificially created lake and it would not work, needs some other checks for FlowableFluid or another proper base class for TR liquids.